cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/MultipleOutputs.java

```diff
@@ -242,6 +245,47 @@ private static Job getNamedJob(JobContext context, String namedOutput) throws IO
     return job;
   }
 
+  /**
+   * Wraps RecordWriter to increment output counters.
```

Comment:
Normally, the user calls `context#write(key, value)` (context in this case is hadoop class), which increments the counter as well as writing the record further in the pipeline.
In the case of multiple outputs, the user calls `context#write(outputName, key, value)` (in this case, the context is CDAP class), and this doesn't at all translate into a call to hadoop's context#write. Because of that, the metrics for output records aren't automatically incremented.

Hadoop's implementation of MultipleOutputs (of which ours is based off of) also does this:
http://grepcode.com/file/repo1.maven.org/maven2/com.ning/metrics.action/0.2.7/org/apache/hadoop/mapreduce/lib/output/MultipleOutputs.java#MultipleOutputs.RecordWriterWithCounter